[GEOS-11390] Replace TestWfsPost with Javascript Demo Page

davidblasby opened the pull request:
[![GEOS-11390](https://badgen.net/badge/JIRA/GEOS-11390/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-11390) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

<!--Include a few sentences describing the overall goals for this Pull Request-->
  
This PR removes the TestWfsPost servlet and replaces it with 100% javascript functionality (running on the browser).

1.  I've included some MIT-licensed JS code.
        SEE: demo-requests.js Line 211 to 277
2. See screenshots, below
     Trying to incorporate andrea's and peter's feedback.
     * button order change on WPS/WCS page to be consistent with Demo Request Page
     * changed button text to be more consistent "<verb> in Request Builder" "<verb> in new page" (verb=Execute, Get Coverage, or Show Result)
4. I will add documentation (new screen shots) once everyone agrees on UX (see screenshots)

Demo Page:
![image](https://github.com/geoserver/geoserver/assets/48937730/4f3a65c5-85e5-4925-916c-a7b19d061170)

WPS:
![image](https://github.com/geoserver/geoserver/assets/48937730/7a66f6c2-90a0-42e3-aa24-7a158077aba0)

WCS:
![image](https://github.com/geoserver/geoserver/assets/48937730/ca48daa1-3fa6-46c6-8fba-5563d8fff9c7)

Demo Requests:
![image](https://github.com/geoserver/geoserver/assets/48937730/e97526ac-36e5-41c4-822c-73fdef0df761)


<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [after feedback] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [n/a] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

jodygarnett commented:
@davidblasby I really like the consistent action button names + workflow.

jodygarnett reviewed (changes requested):
Minor feedback (missing header). When folks are happy with button names we can update the docs and get this moving. 

I would appreciate squash and merge (or interactive rebase) to collapse this to a single commit for backport.

Excellent work David.

petersmythe commented on src/web/demo/src/main/resources/org/geoserver/web/demo/demo-requests.js:
```diff
@@ -0,0 +1,283 @@
+// ======================================================
+// https://www.npmjs.com/package/prettify-xml
+//
+// The MIT License (MIT)
```
There is an MIT License right in the middle of the javascript, is that intended?

davidblasby replied:
Yes - there's MIT code in the JS.  Thats the "xml pretty print" code.

I can move it to its own file??

petersmythe reviewed (approved):
I am happy with the consistent button naming, thank you.

### Additional non blocking feedback from me:

- Clicking Show Result, the Response Body spills over the right edge of the bolded Response Body panel, instead of wrapping.

![image](https://github.com/geoserver/geoserver/assets/5699667/af0252a9-f4dc-421e-b8bc-8a9bed5bb181)

- The (red) Busy spinner icon is not the expected green/black busy standard icon of other GeoServer pages, nor is it in the top right corner, where I expected it to be.

- DemoRequests > Show Result in New Page does not open in a new tab, as expected, but rather in the current tab.  Control/Shift clicking does not change this.

aaime reviewed (changes requested):
Strong recommendation to remove the black boxes around the headers and responses, they are alien to the GeoServer UI, can make the output harder to read.

![image](https://github.com/geoserver/geoserver/assets/325433/8c3106a0-3418-4a41-8395-5c3e314d96ab)

Instead, use horizontal line separators like in the rest of the GUI (same appearance),
they would fit better and not interfere with the output.

![image](https://github.com/geoserver/geoserver/assets/325433/0bbf0d06-6b1d-42ec-9b18-23bbce13b4e8)

In addition, just thinking out loud, is the XML pretty printer able to wrap text at a certain amount of columns? Geometries may otherwise be represented as a very very very long line.

The two other builders look good, but one button label is misleading in both the WCS and WPS pages:

![image](https://github.com/geoserver/geoserver/assets/325433/47207608-b06b-41e3-b04f-6eb9b0cdb625)

There is no such thing as a "Demo Request Builder", had it been a "builder" we would have had no need for the WCS/WPS one. Just say "Get Coverage in Demo requests" and "Execute in Demo requests" instead.

Finally, as @jodygarnett always reminds me, documentation should be updated any time we touch the GUI.

These pages have been affected by the change, screenshots should be re-made, and descriptive text updated accordingly:

* https://docs.geoserver.org/latest/en/user/configuration/demos/index.html
* https://docs.geoserver.org/latest/en/user/services/wcs/requestbuilder.html
* https://docs.geoserver.org/latest/en/user/services/wps/requestbuilder.html

jodygarnett commented:
> * DemoRequests > Show Result in New Page does not open in a new tab, as expected, but rather in the current tab.  Control/Shift clicking does not change this.

jodygarnett commented:
Sorry I accidentally closed, was trying to take it out of draft and add a checklist

davidblasby commented:
@aaime 

1.  Added "standard" GS Spinner


2. `DemoRequests > Show Result in New Page does not open in a new tab, as expected, but rather in the current tab. Control/Shift clicking does not change this.` - didn't do anything on this.  Its doing a POST, so control/shift clicking will not work (its not a link).  In the future, we can probably set the form to `target="_blank"`.  However, I like the fact it opens the requests in its own page.


3. I've moved the "main" javascript into two files - demo-requests.js and xml-pretty-print.js
   However, there's a few lines left over;
   a) on the elements (click).  This is probably removable by adding the click handler with something in the demo-requests.js, but I'm not sure with the wicket page-lifestyle if it will be correctly handled.
   b) in the Wicket AJAX callback.  
   
   I'm not sure how to address this - I'm not sure if its possible.  Wicket puts JS in the page - and definitely in the ajax callbacks (search the GS codebase for `appendJavaScript`).    The Wicket page-lifestyle model makes this difficult as well.
   
   I've done what I can - not sure how to move forward on this.  Like I said, I'm not sure if its possible (look at the source for a wicket page and you will see a bunch of inline script tags) - but perhaps someone with a lot more wicket knowledge than I can move this forward.

5. I removed the headers/response border and used the standard <legend> to do a horizontal line
6. I changed the output so it does a word-wrap.  Not 100% sure if I like this (i.e. a big geometry now takes up a lot of vertical space) or not, but its in and I think useful.
7.  changed to "Demo Requests" (from Demo Requests Builder).

I actually went back to the old Demo Requests - and the new one is very much an improvement.

![image](https://github.com/geoserver/geoserver/assets/48937730/abdfb67a-85b2-4ec9-aff1-bd8654527d42)

jodygarnett commented on doc/en/user/source/services/wps/requestbuilder.rst:
```diff
@@ -1 +1,68 @@



 
```
Did this really have windows line formatting before? Wow ...

jodygarnett reviewed (approved):
Thanks for adding documentation and addressing feedback.

jodygarnett commented:
@aaime if I understand correctly David is at the limits of what he can do with respect to inline javascript and CSP restrictions. Reading wicket 9 guidance the result is to restrict, but a setting can be defined per page to allow pages like this to function.

Reference:
- https://stackoverflow.com/questions/73783715/using-csp-rules-per-page

aaime commented:
@jodygarnett I was just pointing out the conflict with @sikeoka work, if this gets merged as is, I guess Steve is going (unhappily maybe) to pick it up and move the JS out of the HTML pages.

sikeoka commented:
I actually don't care about inline JavaScript in Wicket pages. Wicket 9 has its own complex and strict way of handling Content-Security-Policy and my work just assumes that it will cover everything else in GeoServer while letting Wicket do its own thing.

jodygarnett commented:
I see @sikeoka has now commented.

Here was my research:

- 63 files: https://github.com/search?q=repo%3Ageoserver%2Fgeoserver+script+language%3Ahtml&type=code
- 62 files: https://github.com/search?q=repo%3Abradh%2Fgeoserver+script+language%3Ahtml&type=code

I do not know how to search just on the https://github.com/bradh/geoserver/tree/wicket8 branch.

jodygarnett commented:
@sikeoka perhaps you could look at this PR; it would make us feel a lot better if someone with your wicket experience is comfortable with the change.

jodygarnett commented:
@aaime I think with @sikeoka 's comment all your feedback has been addressed.

aaime commented:
Yes yes, my comment was meant to catch Steve's attention, not to raise a blocker.

aaime commented:
The backport to `2.24.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit 2df22c7ac7c445914ec737c447f87652454d2adf is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-7672-to-2.24.x 65bf9f4e60] WIP: remove Demo Request proxy server - TestWFSPost
 Author: david.blasby <david.blasby@geocat.net>
 Date: Wed May 15 12:17:20 2024 -0700
 1 file changed, 199 insertions(+), 1 deletion(-)
[backport-7672-to-2.24.x 658f570168] wip: wicket side stuff
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 09:11:21 2024 -0700
 1 file changed, 1 insertion(+), 1 deletion(-)
[backport-7672-to-2.24.x 1fc08aeb5d] wip: more wicket side stuff
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 09:12:04 2024 -0700
 4 files changed, 57 insertions(+), 48 deletions(-)
[backport-7672-to-2.24.x 8b4110e8a7] wip
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 12:38:11 2024 -0700
 5 files changed, 119 insertions(+), 14 deletions(-)
[backport-7672-to-2.24.x 07f345d43e] update WCS request page to piggy back on top of DemoRequests page
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 13:09:53 2024 -0700
 2 files changed, 21 insertions(+), 3 deletions(-)
[backport-7672-to-2.24.x 65ff5b8aef] handle WPS request builder
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 13:24:04 2024 -0700
 2 files changed, 17 insertions(+), 5 deletions(-)
[backport-7672-to-2.24.x 259e83dc25] wip: remove demo request, fix test cases, add new test cases
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 16:59:22 2024 -0700
 10 files changed, 121 insertions(+), 1009 deletions(-)
 delete mode 100644 src/wfs/src/main/java/org/vfny/geoserver/wfs/servlets/TestWfsPost.java
 delete mode 100644 src/wfs/src/test/java/org/vfny/geoserver/wfs/servlets/TestWfsPostOnlineIntegrationTest.java
 delete mode 100644 src/wfs/src/test/java/org/vfny/geoserver/wfs/servlets/TestWfsPostTest.java
[backport-7672-to-2.24.x 3acd2f3154] feeback from peter and andrea
 Author: david.blasby <david.blasby@geocat.net>
 Date: Tue May 21 11:22:39 2024 -0700
 14 files changed, 514 insertions(+), 291 deletions(-)
 create mode 100644 src/web/demo/src/main/resources/org/geoserver/web/demo/demo-requests.js

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.24.x 2.24.x
# Navigate to the new working tree
cd .worktrees/backport-2.24.x
# Create a new branch
git switch --create backport-7672-to-2.24.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 4a4aa7d5f3d9fe576dba8a960f28d7e1dcc68fe0,9bafb2e0a6192929aebfad1321fed6d0fb53bcc5,b5c1018e8b2cd60441005905d9f5370bf51ea525,d17606d945830070bfce04a33f5ec314643ea930,ad0eee1c2db1d1d10f8dcb3076aac75b9eb92b72,cad932a5e0430148205636d72045caff78163ea6,4a31b7f56c2c20ee2d3f27f93ea64c1f502d737e,7c86a4cfb77b93bfd9494ca196cf365cd1676c84,2df22c7ac7c445914ec737c447f87652454d2adf,5019d4ead60e69e391ab0495bacbb22ec7d4532f,c54952edc7f370289d2c80d33f29c93a1dd8c2ec,861fecc13582ac2ac7cde6dc367a50f29a7ca6a5,6228298b0dab7c9331aa1567b332cca7d10776c9
# Push it to GitHub
git push --set-upstream origin backport-7672-to-2.24.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.24.x
```
Then, create a pull request where the `base` branch is `2.24.x` and the `compare`/`head` branch is `backport-7672-to-2.24.x`.

aaime commented:
The backport to `2.25.x` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit 2df22c7ac7c445914ec737c447f87652454d2adf is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
[backport-7672-to-2.25.x b668ec27fc] WIP: remove Demo Request proxy server - TestWFSPost
 Author: david.blasby <david.blasby@geocat.net>
 Date: Wed May 15 12:17:20 2024 -0700
 1 file changed, 199 insertions(+), 1 deletion(-)
[backport-7672-to-2.25.x de046b032e] wip: wicket side stuff
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 09:11:21 2024 -0700
 1 file changed, 1 insertion(+), 1 deletion(-)
[backport-7672-to-2.25.x c25f279e8d] wip: more wicket side stuff
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 09:12:04 2024 -0700
 4 files changed, 57 insertions(+), 48 deletions(-)
[backport-7672-to-2.25.x 7583e5971c] wip
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 12:38:11 2024 -0700
 5 files changed, 119 insertions(+), 14 deletions(-)
[backport-7672-to-2.25.x d076f9591e] update WCS request page to piggy back on top of DemoRequests page
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 13:09:53 2024 -0700
 2 files changed, 21 insertions(+), 3 deletions(-)
[backport-7672-to-2.25.x 5c4d2c7055] handle WPS request builder
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 13:24:04 2024 -0700
 2 files changed, 17 insertions(+), 5 deletions(-)
[backport-7672-to-2.25.x f9e46a2bd2] wip: remove demo request, fix test cases, add new test cases
 Author: david.blasby <david.blasby@geocat.net>
 Date: Thu May 16 16:59:22 2024 -0700
 10 files changed, 121 insertions(+), 1009 deletions(-)
 delete mode 100644 src/wfs/src/main/java/org/vfny/geoserver/wfs/servlets/TestWfsPost.java
 delete mode 100644 src/wfs/src/test/java/org/vfny/geoserver/wfs/servlets/TestWfsPostOnlineIntegrationTest.java
 delete mode 100644 src/wfs/src/test/java/org/vfny/geoserver/wfs/servlets/TestWfsPostTest.java
[backport-7672-to-2.25.x 9edb6e4129] feeback from peter and andrea
 Author: david.blasby <david.blasby@geocat.net>
 Date: Tue May 21 11:22:39 2024 -0700
 14 files changed, 514 insertions(+), 291 deletions(-)
 create mode 100644 src/web/demo/src/main/resources/org/geoserver/web/demo/demo-requests.js

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.25.x 2.25.x
# Navigate to the new working tree
cd .worktrees/backport-2.25.x
# Create a new branch
git switch --create backport-7672-to-2.25.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 4a4aa7d5f3d9fe576dba8a960f28d7e1dcc68fe0,9bafb2e0a6192929aebfad1321fed6d0fb53bcc5,b5c1018e8b2cd60441005905d9f5370bf51ea525,d17606d945830070bfce04a33f5ec314643ea930,ad0eee1c2db1d1d10f8dcb3076aac75b9eb92b72,cad932a5e0430148205636d72045caff78163ea6,4a31b7f56c2c20ee2d3f27f93ea64c1f502d737e,7c86a4cfb77b93bfd9494ca196cf365cd1676c84,2df22c7ac7c445914ec737c447f87652454d2adf,5019d4ead60e69e391ab0495bacbb22ec7d4532f,c54952edc7f370289d2c80d33f29c93a1dd8c2ec,861fecc13582ac2ac7cde6dc367a50f29a7ca6a5,6228298b0dab7c9331aa1567b332cca7d10776c9
# Push it to GitHub
git push --set-upstream origin backport-7672-to-2.25.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.25.x
```
Then, create a pull request where the `base` branch is `2.25.x` and the `compare`/`head` branch is `backport-7672-to-2.25.x`.

jodygarnett commented on src/web/demo/src/main/java/org/geoserver/web/demo/DemoRequestsPage.java:
```diff
@@ -63,7 +93,30 @@ public class DemoRequestsPage extends GeoServerBasePage {
+    }
+
+    @Override
+    public void renderHead(IHeaderResponse response) {
```
Re-opening https://osgeo-org.atlassian.net/browse/GEOS-11390 due to missing call to super.render(response) here, required to ensure header contributions are included.
